### PR TITLE
Johnc lmm matrix export

### PR DESCRIPTION
**Pull request by johnc1231**
-Created KinshipMatrix class in Python
-Created method rrm() on VDS that returns a KinshipMatrix
-Modified lmmreg to take a KinshipMatrix instead of a VDS from which to compute kinship. 
-Created tests for KinshipMatrix.

### CODE REVIEW

**Comment by jbloom22** [comment]
ill make a PR on docs, and comments on code, then you can rebase

**Comment by johnc1231** [comment]
Alright cool. I'll hold off on handling the failing test until you do that.

**Comment by johnc1231** [comment]
Fixed the test. Feel free to look over whenever

File: python/hail/dataset.py [hunk 1]
```diff
@@ -3317,6 +3301,43 @@ def repartition(self, num_partitions, shuffle=True):
...
 
     @handle_py4j
+    def rrm(self, force_block = False, force_gramian = False):
+        """Computes the Realized Relationship Matrix based on this dataset.
+
+        **Examples**
+
+        Compute the RRM for `vds`.
+
+        >>> vds.rrm()
```
Comment (tpoterba) [inline python/hail/dataset.py]: store in a variable

File: src/main/scala/is/hail/methods/LinearMixedRegression.scala [hunk 2]
```diff
@@ -27,46 +27,38 @@ object LinearMixedRegression {
...
-      case (true, true) => fatal("Cannot force both Block and Grammian")
-      case (b, _) => b
-    }
-    
-    val filtKinshipVds = kinshipVds.filterSamples((s, sa) => completeSamplesSet(s))
-    if (filtKinshipVds.sampleIds != completeSamples)
+    val filtKinshipMatrix = kinshipMatrix.filterSamples(completeSamplesSet.contains)
+
+    if (filtKinshipMatrix.sampleIds.toVector != completeSamples)
       fatal("Array of sample IDs in assoc_vds and array of sample IDs in kinship_vds (with both filtered to complete samples in assoc_vds) do not agree. This should not happen when kinship_vds is formed by filtering variants on assoc_vds.")
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/LinearMixedRegression.scala]: break this line

File: src/main/scala/is/hail/methods/LinearMixedRegression.scala [hunk 3]
```diff
@@ -76,11 +68,11 @@ object LinearMixedRegression {
...
     info(s"lmmreg: running lmmreg on $n samples with $k sample ${plural(k, "covariate")} including intercept...")
 
-    info(s"lmmreg: Computing RRM for $n samples...")
+    val cols = filtKinshipMatrix.matrix.numCols().toInt
 
-    val (rrm, m) = ComputeRRM(filtKinshipVds, useBlock)
+    val rrm = new DenseMatrix[Double](cols, cols, filtKinshipMatrix.matrix.toBlockMatrix().toLocalMatrix().toArray)
 
-    info(s"lmmreg: RRM computed using $m variants")
+    //info(s"lmmreg: RRM computed using $m variants")
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/LinearMixedRegression.scala]: commented code

**Comment by johnc1231** [comment]
Addressed Tim's comments

**Comment by jbloom22** [comment]
Merge my doc suggestions, rebase, and I'll look again.

**Comment by johnc1231** [comment]
@jbloom22 Added your docs changes and rebased on master, back to you.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package is.hail.methods
+
+import org.apache.spark.mllib.linalg.Vectors
+import org.apache.spark.mllib.linalg.distributed.{IndexedRow, IndexedRowMatrix}
+import scala.collection.Searching._
+
+/**
+  * Represents a KinshipMatrix, which is a matrix where the number at index (i, j) represents how related the samples at index
+  * i and j in the samples array are.
+  */
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: Represents a kinship matrix. Entry (i, j) of the matrix encodes the relatedness of the ith and jth samples in sampleIds.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+class KinshipMatrix(val matrix: IndexedRowMatrix, val sampleIds: Array[String]) {
+  assert(matrix.numCols().toInt == matrix.numRows().toInt && matrix.numCols().toInt == sampleIds.length)
+
+  /**
+    * Filters list of samples based on predicate, and removes corresponding row and column from the list.
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: "rows and columns from the matrix."

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    * @param pred The predicate that decides whether a sample is kept.
+    */
+  def filterSamples(pred: (String => Boolean)): KinshipMatrix = {
+    val (samplePairsToTake, samplePairsToDrop) = sampleIds.zipWithIndex.partition(pair => pair match {case (sampleName, index) => pred(sampleName)})
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: Use Keep rather than Take

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+    val filteredSamples = samplePairsToTake.map(_._1)
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: filteredSampleIds

File: src/main/scala/is/hail/stats/ComputeRRM.scala [hunk 8]
```diff
@@ -1,42 +1,74 @@
...
-      val mRec = 1d / A.numRows()
-      (ComputeLocalGrammian.withoutBlock(A) :* mRec, A.numRows().toInt)
+      rowCount = A.numRows()
+      computedGramian = ComputeGramian.withoutBlock(A)
     }
+
+    val mRec = 1d / rowCount
+
+    (new IndexedRowMatrix(computedGramian.rows.map(ir => new IndexedRow(ir.index, ir.vector.map(_ * mRec)))), rowCount)
+    //val scaledBlockRDD = computedGramian.blocks.map(tuple => tuple match {case (coords, matrix) => (coords, scaleMatrix(matrix, mRec))})
```
Comment (jbloom22) [inline src/main/scala/is/hail/stats/ComputeRRM.scala]: commented code

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 9]
```diff
@@ -927,6 +929,15 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
 
   def sampleQC(root: String = "sa.qc"): VariantDataset = SampleQC(vds, root)
 
+  def rrm(forceBlock : Boolean = false, forceGramian : Boolean = false): KinshipMatrix = {
+    requireSplit("rrm")
+    info(s"rrm: Beginning RRM computation")
```
Comment (jbloom22) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: "rrm: Computing Realized Relationship Matrix..."

File: src/main/scala/is/hail/variant/VariantDataset.scala [hunk 10]
```diff
@@ -927,6 +929,15 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
...
+    val (rrm, m) = ComputeRRM(vds, forceBlock, forceGramian)
+    info(s"rrm: RRM computed using $m variants")
```
Comment (jbloom22) [inline src/main/scala/is/hail/variant/VariantDataset.scala]: add a period

File: src/main/scala/is/hail/methods/LinearMixedRegression.scala [hunk 11]
```diff
@@ -27,45 +27,37 @@ object LinearMixedRegression {
...
-    val useBlock = (forceBlock, forceGrammian) match {
-      case (false, false) => y.length > 3000 // for small matrices, computeGrammian fits in memory and runs faster than BlockMatrix product
-      case (true, true) => fatal("Cannot force both Block and Grammian")
-      case (b, _) => b
-    }
-    
-    val filtKinshipVds = kinshipVds.filterSamples((s, sa) => completeSamplesSet(s))
-    if (filtKinshipVds.sampleIds != completeSamples)
-      fatal("Array of sample IDs in assoc_vds and array of sample IDs in kinship_vds (with both filtered to complete samples in assoc_vds) do not agree. This should not happen when kinship_vds is formed by filtering variants on assoc_vds.")
+    val filtKinshipMatrix = kinshipMatrix.filterSamples(completeSamplesSet)
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/LinearMixedRegression.scala]: change all my "filt" to "filtered" for consistency

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    val sampleNumsToDropArray = samplePairsToDrop.map(_._2)
+    val sampleNumsToDropSet = sampleNumsToDropArray.toSet
+    val sampleNumsToTakeArray = samplePairsToTake.map(_._2)
+
+    val filteredRows = matrix.rows.filter(ir => !sampleNumsToDropSet(ir.index.toInt))
+    val filteredRowsAndCols = filteredRows.map(ir => {
+      val numBelowToDelete = sampleNumsToDropArray.search(ir.index.toInt) match {case InsertionPoint(i) => i}
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: matching on only one case can be done with:
`val InsertionPoint(numBelowToDelete) = sampleNumsToDropArray.search(ir.index.toInt)`

File: src/main/scala/is/hail/methods/KinshipMatrix.scala (hunk 6 above)
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: I find samplePairs very confusing in this kinship context, when it's really sampleID and index. How about `sampleWithIndex`? Then using `pred(pair._1)` in place of `pair match {case (sampleName, index) => pred(sampleName)}` will be clearer.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+
+  /**
+    * Filters list of samples based on predicate, and removes corresponding row and column from the list.
+    * @param pred The predicate that decides whether a sample is kept.
+    */
+  def filterSamples(pred: (String => Boolean)): KinshipMatrix = {
+    val (samplePairsToTake, samplePairsToDrop) = sampleIds.zipWithIndex.partition(pair => pair match {case (sampleName, index) => pred(sampleName)})
+
+    val filteredSamples = samplePairsToTake.map(_._1)
+    val sampleNumsToDropArray = samplePairsToDrop.map(_._2)
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: I prefer Indices to Nums.  Num is usually a count.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    val sampleNumsToDropSet = sampleNumsToDropArray.toSet
+    val sampleNumsToTakeArray = samplePairsToTake.map(_._2)
+
+    val filteredRows = matrix.rows.filter(ir => !sampleNumsToDropSet(ir.index.toInt))
+    val filteredRowsAndCols = filteredRows.map(ir => {
+      val numBelowToDelete = sampleNumsToDropArray.search(ir.index.toInt) match {case InsertionPoint(i) => i}
+      val index = ir.index - numBelowToDelete
+      val arrayVec = ir.vector.toArray
```
Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: for consistency: `vecArray`

Comment (jbloom22) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: Can you avoid all this additional full copy in typical case that the array backing the vector is the array you get from toArray?

File: src/main/scala/is/hail/stats/ComputeRRM.scala [hunk 15]
```diff
@@ -1,42 +1,74 @@
...
+    //(new BlockMatrix(scaledBlockRDD, computedGramian.rowsPerBlock, computedGramian.colsPerBlock), rowCount.toInt)
+  }
+}
+
+object LocalDenseMatrixToIndexedRowMatrix {
+  def apply(dm: DenseMatrix[Double], sc: SparkContext): IndexedRowMatrix = {
+    //TODO Is there a better Breeze to Spark conversion?
+    val range = (0 until dm.rows).par
+    val numberedDVs: ParSeq[IndexedRow] = range.map(rowNum => IndexedRow(rowNum.toLong, (dm(rowNum, ::).t)))
```
Comment (jbloom22) [inline src/main/scala/is/hail/stats/ComputeRRM.scala]: can you explain what's going on with parallel seq here?

Comment (johnc1231) [inline src/main/scala/is/hail/stats/ComputeRRM.scala]: Shouldn't be there, sorry. Was playing around with something, thought that I dropped that out.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala (hunk 12 above)
Comment (johnc1231) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: That's really cool. Didn't realize that worked. Thanks.

File: src/main/scala/is/hail/methods/KinshipMatrix.scala (hunk 14 above)
Comment (johnc1231) [inline src/main/scala/is/hail/methods/KinshipMatrix.scala]: Can you elaborate here? Where is the full copy? DenseVector.toArray doesn't copy anything. And Vectors.dense isn't copying anything either.

**Comment by johnc1231** [comment]
Conflicts handled, back to you @jbloom22

**Comment by jbloom22** [comment]
rebase

**Comment by johnc1231** [comment]
@jbloom22 Rebased, back to you.

**Pull request merged**